src/libstore/derivations.cc

```diff
@@ -272,7 +273,17 @@ Derivation parseDerivation(const Store & store, std::string && s, std::string_vi
 
 static void printString(string & res, std::string_view s)
 {
-    char buf[s.size() * 2 + 2];
+    // Large stack allocations can skip past the stack protection page.
+    const size_t stack_protection_size = 4096;
+    // We reduce the max stack allocated buffer by an extra amount to increase
+    // the chance of hitting it, even when `fun`'s first access is some distance
+    // into its *further* stack frame, particularly if the call was inlined and
+    // therefore not writing a frame pointer.
+    const size_t play = 64 * sizeof(char *); // 512B on 64b archs
+
+    boost::container::small_vector<char, stack_protection_size - play> buffer;
```

Comment:
```suggestion
    boost::container::small_vector<char, 64 * 1024> buffer;
```
I don't think we need to worry too much about overflows for non-recursive functions. But if we're worried about this, it's probably better to use gcc's `-fstack-check` flag, which enables stack probe instructions automatically.

Reply:
> I don't think we need to worry too much about overflows for non-recursive functions

_Some_ limit on stack usage is necessary though, because derivation attributes can be of disproportional size; multiple megabytes.

Considering the distribution of sizes, it really doesn't matter how big you go beyond the first page. Probably well over 95% already fits anyway. The benefit of staying within a page size is not needing `-fstack-check` in this location.

`-fstack-check` is definitely a good idea for improving the correct functioning of stack overflow detection and memory safety. We'll need to know the overhead though, and what to do for llc. Can we please scope this out?

I'd *really* like to move on with this and I hope you do too. If we really need a fourth iteration on this simple change, just tell me exactly what to do so we can be done with it.

Reply:
> just tell me exactly what to do so we can be done with it.

I ded provide a code suggestion, namely use 64 KB instead of a complicated calculation involving page sizes and char * sizes.

Reply:
Alright, apparently I was focusing way too much on the regular comments. Thanks for being patient :)
I've committed it. Good to go now?